lib/util/cat.js

```diff
@@ -5,51 +5,272 @@
+      tx.id, // TransactionGuid
+      false // force a transaction trace to be recorded
+    ])
+  } catch (err) {
+    logger.trace(
+      err,
+      'Failed to serialize transaction: %s - not adding CAT response headers',
+      txName
+    )
+    return
+  }
+
+  const encKey = config.encoding_key
+  const obfAppData = hashes.obfuscateNameUsingKey(appData, encKey)
+  return { key: HTTP_CAT_APP_DATA_HEADER, data: obfAppData }
+}
+
+/**
+ * Adds CAT headers to outbound request.
+ *
+ * @param {Object} config agent config
+ * @param {Transaction} tx
+ * @param {Object} headers object that contains headers the agent is adding to client request
+ */
+cat.addCatHeaders = function addCatHeaders(config, tx, headers, useMqHeaders) {
+  if (!config.encoding_key) {
+    logger.warn('Missing encoding key, not adding CAT headers!')
+    return
+  }
+
+  const idHeader = useMqHeaders ? MQ_CAT_ID_HEADER : HTTP_CAT_ID_HEADER
+  const txHeader = useMqHeaders ? MQ_CAT_TRANSACTION_HEADER : HTTP_CAT_TRANSACTION_HEADER
+
+  // Add in the application ID
+  if (config.obfuscatedId) {
+    headers[idHeader] = config.obfuscatedId
+  }
+
+  const txName = tx.getFullName() || ''
+
+  const pathHash = hashes.calculatePathHash(config.applications()[0], txName, tx.referringPathHash)
+  tx.pushPathHash(pathHash)
+
+  try {
+    const txData = hashes.obfuscateNameUsingKey(
+      JSON.stringify([tx.id, false, tx.tripId || tx.id, pathHash]),
+      config.encoding_key
+    )
+    headers[txHeader] = txData
+
+    logger.trace('Added outbound request CAT headers in transaction %s', tx.id)
+  } catch (err) {
+    logger.trace(err, 'Failed to create CAT payload')
+  }
+}
+
+/**
+ * Find the CAT id, transaction, app data headers
+ * from the headers of either HTTP or MQ request
+ *
+ * @param {Object} headers
+ * @return {Object} { id, transactionId, appData }
+ */
+cat.extractCatHeaders = function extractCatHeaders(headers) {
+  // Hunt down the CAT headers.
+  let id = null
+  let transactionId = null
+  let appData = null
+  // eslint-disable-next-line guard-for-in
+  for (const key in headers) {
+    if (MATCH_CAT_ID_HEADER.test(key)) {
+      id = headers[key]
+    } else if (MATCH_CAT_TRANSACTION_HEADER.test(key)) {
+      transactionId = headers[key]
+    } else if (MATCH_CAT_APP_DATA_HEADER.test(key)) {
+      appData = headers[key]
+    }
+    if (id && transactionId && appData) {
+      break
+    }
+  }
+
+  return { id, transactionId, appData }
+}
+
+/**
+ * Extracts the account Id from CAT data and verifies if it is
+ * a trusted account id
+ *
+ * @param {Object} CAT data
+ * @param {Array} trustedAccounts from config
+ * @return {Boolean}
+ */
+cat.isTrustedAccountId = function isTrustedAccountId(data, trustedAccounts) {
+  const accountId = parseInt(data.split('#')[0], 10)
+  const trusted = trustedAccounts.includes(accountId)
+  if (!trusted) {
+    logger.trace('Response from untrusted CAT header account id: %s', accountId)
+  }
```

Comment:
Feels a little odd to be logging from a logic check function VS where checked but this will make cleanup pretty easy so makes sense to leave it here.

Reply:
yea, this was a tough one because there was logging and it's based on the parsed value